File: 318.maximum-product-of-word-lengths.py
#
# @lc app=leetcode id=318 lang=python3
#
# [318] Maximum Product of Word Lengths
#

# Time:  O(n^2)
# Space: O(n)

from typing import List


# @lc code=start
class Solution:
    def maxProduct(self, words: List[str]) -> int:
        res = 0
        mask = [0] * len(words)
        for i in range(len(words)):
            for c in words[i]:
                mask[i] |= 1 << (ord(c) - 97)
            for j in range(i):
                if not (mask[i] & mask[j]):
                    res = max(res, len(words[i]) * len(words[j]))
        return res


# @lc code=end

File: 318.maximum-product-of-word-lengths.js
/*
 * @lc app=leetcode id=318 lang=javascript
 *
 * [318] Maximum Product of Word Lengths
 */

// @lc code=start
/**
 * @param {string[]} words
 * @return {number}
 */
var maxProduct = function (words) {
  let res = 0;
  const mask = Array(words.length).fill(0);
  for (let i = 0; i < words.length; ++i) {
    for (let j = 0; j < words[i].length; ++j) {
      mask[i] |= 1 << (words[i][j].charCodeAt() - 97);
    }
    for (let j = 0; j < i; ++j) {
      if (!(mask[i] & mask[j])) {
        res = Math.max(res, words[i].length * words[j].length);
      }
    }
  }
  return res;
};
// @lc code=end
